File: clientlib/python/scan_wizard.py
#!/usr/bin/env python3

# Scan Wizard application.
#
# This program starts scanning of new Flic buttons that have not previously been verified by the server.
# Once it finds a button that is in private mode, it shows a message that the user should hold it down for 7 seconds to make it public.
# Once it finds a button that is in public mode, it attempts to connect to it.
# If it could be successfully connected and verified, the bluetooth address is printed and the program exits.
# If it could not be verified within 30 seconds, the scan is restarted.

import fliclib

client = fliclib.FlicClient("localhost")

def done(bd_addr):
	print("Button " + bd_addr + " was successfully added!")
	client.close()

def on_adv_packet(scanner, bd_addr, name, rssi, is_private, already_verified):
	if already_verified:
		return
	
	if is_private:
		print("Button " + bd_addr + " is currently private. Hold it down for 7 seconds to make it public.")
		return
	
	print("Found public button " + bd_addr + ", now connecting...")
	client.remove_scanner(scanner)
	
	def restart_scan():
		print("Restarting scan")
		client.add_scanner(scanner)
	
	def on_create(channel, error, connection_status):
		if connection_status == fliclib.ConnectionStatus.Ready:
			done(bd_addr)
		elif error != fliclib.CreateConnectionChannelError.NoError:
			print("Failed: " + str(error))
			restart_scan()
		else:
			client.set_timer(30 * 1000, lambda: client.remove_connection_channel(channel))
	
	def on_removed(channel, removed_reason):
		print("Failed: " + str(removed_reason))
		restart_scan()
	
	def on_connection_status_changed(channel, connection_status, disconnect_reason):
		if connection_status == fliclib.ConnectionStatus.Ready:
			done(bd_addr)
	
	channel = fliclib.ButtonConnectionChannel(bd_addr)
	channel.on_create_connection_channel_response = on_create
	channel.on_removed = on_removed
	channel.on_connection_status_changed = on_connection_status_changed
	client.add_connection_channel(channel)
	
scanner = fliclib.ButtonScanner()
scanner.on_advertisement_packet = on_adv_packet
client.add_scanner(scanner)

client.handle_events()

File: clientlib/nodejs/scanwizard.js
/*
 * Use this tool to pair one Flic button.
 * Run the example.js to get button events.
 */

var fliclib = require("./fliclibNodeJs");
var FlicClient = fliclib.FlicClient;
var FlicConnectionChannel = fliclib.FlicConnectionChannel;
var FlicScanner = fliclib.FlicScanner;

var client = new FlicClient("localhost", 5551);

function startScanWizard() {
	console.log("Welcome to the add new button wizard. Press and hold down your Flic button to add it.");
	
	var timeout = null;
	
	function success(cc, bdAddr) {
		client.removeConnectionChannel(cc);
		console.log("Button " + bdAddr + " successfully added!");
		done();
	}
	function failed(msg) {
		console.log("Scan Wizard Failed: " + msg);
		done();
	}
	function done() {
		client.close();
		if (timeout != null) {
			clearTimeout(timeout);
		}
	}
	var cc = null;
	var scanner = new FlicScanner();
	scanner.on("advertisementPacket", function(bdAddr, name, rssi, isPrivate, alreadyVerified, alreadyConnectedToThisDevice, alreadyConnectedToOtherDevice) {
		if (alreadyVerified) {
			return;
		}
		if (isPrivate) {
			console.log("Your button is private. Hold down for 7 seconds to make it public.");
			return;
		}
		client.removeScanner(scanner);
		
		cc = new FlicConnectionChannel(bdAddr);
		cc.on("createResponse", function(error, connectionStatus) {
			if (connectionStatus == "Ready") {
				// Got verified by someone else between scan result and this event
				success(cc, bdAddr);
			} else if (error != "NoError") {
				failed("Too many pending connections");
			} else {
				console.log("Found a public button. Now connecting...");
				timeout = setTimeout(function() {
					client.removeConnectionChannel(cc);
				}, 30 * 1000);
			}
		});
		cc.on("connectionStatusChanged", function(connectionStatus, disconnectReason) {
			if (connectionStatus == "Ready") {
				success(cc, bdAddr);
			}
		});
		cc.on("removed", function(removedReason) {
			if (removedReason == "RemovedByThisClient") {
				failed("Timed out");
			} else {
				failed(removedReason);
			}
		});
		client.addConnectionChannel(cc);
	});
	client.addScanner(scanner);
}

client.on("ready", startScanWizard);
